File: importers/health_contract.py
"""Pure, versioned health-entry contract shared by importers."""

from __future__ import annotations

import re
from datetime import datetime
from typing import Any

CONTRACT_VERSION = "health.v1"
ENTRY_TYPES = frozenset({"steps", "weight", "heart_rate", "sleep", "exercise", "body_composition", "measurement_goal"})
ALIASES = {
    "heartRate": "heart_rate", "heart-rate": "heart_rate", "hr": "heart_rate",
    "bodyComp": "body_composition", "body-comp": "body_composition",
    "goal": "measurement_goal", "measurement-goal": "measurement_goal",
}
_FIELDS = {
    "steps": ("count", ("steps",)),
    "weight": ("weight_kg", ("weight",)),
    "heart_rate": ("bpm", ("beats_per_minute", "heart_rate")),
    "sleep": ("duration_hours", ("duration_h",)),
    "exercise": ("duration_s", ("duration_seconds",)),
}
_UUID = re.compile(r"^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-5][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$")


def _number(value: Any) -> bool:
    return isinstance(value, (int, float)) and not isinstance(value, bool)


def _timestamp(value: Any) -> str:
    if not isinstance(value, str) or not value or value.endswith("Z") is False and "+" not in value and not re.search(r"T[^ ]*-[0-9]{2}:[0-9]{2}$", value):
        raise ValueError("timestamp must be an ISO-8601 string with an explicit offset")
    try:
        datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError as exc:
        raise ValueError("timestamp must be an ISO-8601 string with an explicit offset") from exc
    return value


def identity(record: dict[str, Any]) -> str:
    source = record.get("source")
    stable = record.get("source_record_id") or record.get("external_id")
    if stable:
        return f"{source or 'unknown'}:{stable}"
    start = record.get("start") or record.get("timestamp")
    end = record.get("end") or record.get("timestamp")
    if not source or not start or not end:
        raise ValueError("identity requires source record id or provenance/source and start/end")
    return f"fallback:{source}:{start}:{end}:{record.get('entry_type')}"


def validate_health_entry(record: Any) -> dict[str, Any]:
    if not isinstance(record, dict):
        raise ValueError("health entry must be an object")
    entry_type = ALIASES.get(record.get("entry_type"), record.get("entry_type"))
    if entry_type not in ENTRY_TYPES:
        raise ValueError("unsupported entry_type")
    timestamp = _timestamp(record.get("timestamp"))
    value = record.get("value")
    if value is not None and not isinstance(value, dict):
        raise ValueError("value must be an object or null")
    if record.get("numeric_value") is not None and not _number(record["numeric_value"]):
        raise ValueError("numeric_value must be a number or null")
    if record.get("duration_s") is not None and (not _number(record["duration_s"]) or record["duration_s"] < 0):
        raise ValueError("duration_s must be non-negative")
    if record.get("source") is not None and not isinstance(record["source"], str):
        raise ValueError("source must be a string")
    return {"entry_type": entry_type, "timestamp": timestamp, "value": value, "numeric_value": record.get("numeric_value"), "duration_s": record.get("duration_s"), "source": record.get("source"), "external_id": record.get("external_id")}


def normalize_health_entry(record: dict[str, Any]) -> dict[str, Any]:
    checked = validate_health_entry(record)
    value = dict(checked["value"] or {})
    entry_type = checked["entry_type"]
    canonical, aliases = _FIELDS.get(entry_type, (None, ()))
    if canonical and canonical not in value:
        for alias in aliases:
            if alias in value:
                value[canonical] = value[alias]
                break
    if canonical and canonical in value and not _number(value[canonical]):
        raise ValueError(f"{canonical} must be a number")
    if entry_type == "steps" and value.get("count") is not None and value["count"] < 0:
        raise ValueError("steps count must be non-negative")
    result = dict(checked)
    result["value"] = value
    result["identity"] = identity({**record, **result})
    result["contract_version"] = CONTRACT_VERSION
    result["raw_content"] = {"untrusted": True, "value": record}
    return result


def project_health_entries(records: list[dict[str, Any]]) -> list[dict[str, Any]]:
    projected: dict[str, dict[str, Any]] = {}
    for record in records:
        item = normalize_health_entry(record)
        key = item["identity"]
        if key in projected:
            if projected[key] != item:
                raise ValueError(f"conflicting duplicate identity: {key}")
            continue
        projected[key] = item
    return [projected[key] for key in sorted(projected)]

File: supabase/functions/alexandria/health_contract.ts
export const CONTRACT_VERSION = "health.v1" as const;
export const ENTRY_TYPES = new Set([
  "steps", "weight", "heart_rate", "sleep", "exercise", "body_composition", "measurement_goal",
]);
const ALIASES: Record<string, string> = { heartRate: "heart_rate", "heart-rate": "heart_rate", hr: "heart_rate", bodyComp: "body_composition", "body-comp": "body_composition", goal: "measurement_goal", "measurement-goal": "measurement_goal" };
const FIELDS: Record<string, [string, string[]]> = {
  steps: ["count", ["steps"]], weight: ["weight_kg", ["weight"]], heart_rate: ["bpm", ["beats_per_minute", "heart_rate"]], sleep: ["duration_hours", ["duration_h"]], exercise: ["duration_s", ["duration_seconds"]],
};
const number = (value: unknown): value is number => typeof value === "number" && Number.isFinite(value);
const isStr = (value: unknown): value is string => typeof value === "string";
const timestamp = (value: unknown): string => {
  if (typeof value !== "string" || !value || !/[Zz]|[+-]\d\d:\d\d$/.test(value) || Number.isNaN(Date.parse(value))) throw new Error("timestamp must be ISO-8601 with explicit offset");
  return value;
};
export function identity(record: Record<string, unknown>): string {
  const source = record.source as string | undefined;
  const stable = record.source_record_id || record.external_id;
  if (stable) return `${source || "unknown"}:${String(stable)}`;
  const start = record.start || record.timestamp;
  const end = record.end || record.timestamp;
  if (!source || !start || !end) throw new Error("identity requires source record id or provenance/source and start/end");
  return `fallback:${source}:${start}:${end}:${record.entry_type}`;
}
export function validateHealthEntry(record: unknown): Record<string, unknown> {
  if (!record || typeof record !== "object" || Array.isArray(record)) throw new Error("health entry must be an object");
  const input = record as Record<string, unknown>;
  const entryType = ALIASES[String(input.entry_type)] || input.entry_type;
  if (typeof entryType !== "string" || !ENTRY_TYPES.has(entryType)) throw new Error("unsupported entry_type");
  const value = input.value;
  if (value !== undefined && value !== null && (typeof value !== "object" || Array.isArray(value))) throw new Error("value must be an object or null");
  if (input.numeric_value !== undefined && input.numeric_value !== null && !number(input.numeric_value)) throw new Error("numeric_value must be a number or null");
  if (input.duration_s !== undefined && input.duration_s !== null && (!number(input.duration_s) || input.duration_s < 0)) throw new Error("duration_s must be non-negative");
  if (input.source !== undefined && input.source !== null && !isStr(input.source)) throw new Error("source must be a string");
  return { entry_type: entryType, timestamp: timestamp(input.timestamp), value: value ?? null, numeric_value: input.numeric_value ?? null, duration_s: input.duration_s ?? null, source: input.source ?? null, external_id: input.external_id ?? null };
}
export function normalizeHealthEntry(record: Record<string, unknown>): Record<string, unknown> {
  const checked = validateHealthEntry(record);
  const value = { ...((checked.value as Record<string, unknown> | null) || {}) };
  const field = FIELDS[checked.entry_type as string];
  if (field && value[field[0]] === undefined) for (const alias of field[1]) if (value[alias] !== undefined) { value[field[0]] = value[alias]; break; }
  if (field && value[field[0]] !== undefined && !number(value[field[0]])) throw new Error(`${field[0]} must be a number`);
  if (checked.entry_type === "steps" && value.count !== undefined && (value.count as number) < 0) throw new Error("steps count must be non-negative");
  const result = { ...checked, value, identity: identity({ ...record, ...checked }), contract_version: CONTRACT_VERSION, raw_content: { untrusted: true, value: record } };
  return result;
}
export function projectHealthEntries(records: Record<string, unknown>[]): Record<string, unknown>[] {
  const projected = new Map<string, Record<string, unknown>>();
  for (const record of records) { const item = normalizeHealthEntry(record); const key = item.identity as string; const prior = projected.get(key); if (prior) { if (JSON.stringify(prior) !== JSON.stringify(item)) throw new Error(`conflicting duplicate identity: ${key}`); } else projected.set(key, item); }
  return [...projected.entries()].sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0)).map(([, value]) => value);
}
